DEVELOPING.md

```diff
@@ -0,0 +1,53 @@
+
+One key piece of information that is kept with each file is a version number (usually in the header section e.g. `<OpenSimDocument Version="NNNNN">`. The same is true for .sto files that contain a header as well, though we will not discuss .sto files here. The version number encodes what the code was expecting the format to be when the file was produced/written. Keeping this number consistent with the content of the file is the fundamental invariant that needs to be maintained. In code, the version number is maintained in XMLDocument.cpp and is updated as version numbers are incremented. The actual numbers do not matter much, as long as they are monotonically increasing. OpenSim 3.2 shipped with OpenSimDocument Version number 30000. OpenSim 4.0 development started at version 30500 and will ship with 40000).
+
+Deserialization of Objects and their Properties (primitives that are read-from/written-to files) is mostly transparent to users/developers. Once you use the property macros (e.g. ` OpenSim_DECLARE_UNNAMED_PROPERTY`) in the header and perform the necessary wiring/calls, these properties know how to serialize themselves to XML elements and back. This serialization/deserialization, however, assumes a fixed XML layout/schema. If this assumption fails (because of a property name change, layout change or addition of new properties) then deserialization will not work out of the box and the developer making the change is responsible for changing the version number and the deserialization code to be in sync with the latest code. This is very important since if the developer does not update the version number then the code and the XML files will not be in sync and it will be very hard to go back to find out what changes should have been made. 
+
+Keep in mind that deserialization methods are practically called during the construction of the top level object/Model so typically you cannot rely on having any member variables or pointers populated unless done by default constructors, instead deserialization methods operate on the XML structure only. This helps isolate the backward compatibility code from the rest of the modeling/simulation or initialization functions.
+
+Deserialization works by making the following calls:
+ 1. Parsing the XML files (done by the framework, only add a constructor from file if needed)
+ 2. The framework finds an XML tag that corresponds to a subclass of OpenSim::Object, then
```

Comment:
I think 1. and 2. could be a little clearer. What does "framework" mean? I don't think the meaning of "constructor from file" is obvious.

For 1., even if the parsing is done by the framework, isn't some constructor being called with a file name? Maybe 1. could be "A user constructs an object that is a subclass OpenSim::Object by using a constructor that takes the filename of a .osim or .xml file (only some classes have such a constructor)."

Then there could be a new entry (1.5) "The constructor invokes code that parses the XML file."

Also, for 2., is it more correct to say "The (something) iterates over each XML tag, some of which will correspond to subclasses of OpenSim::Object"?